### DISCUSSION TITLE
Functionality to copy trained Spacy NER + Parser model from Spacy2.x to Spacy3.x

### DISCUSSION THREAD

**Discussion by AndriyMulyar** [explosion/spaCy#6330] [Help: Coding & Implementations] [answered]
## Feature description

Add the ability to copy model weights from spacy 2.x model to spacy 3.x model.
Currently there is no way to use a trained NER or parser model from spacy 2.x in spacy3.x. This will likely be a large, discouraging pain point for organizations attempting to move existing code into the 3.x (training data may be gone, inability to dedicate dev hours for something that doesn't add a new feature, etc).

Functionality this would require:
1. Copying encoder + (ner/dependency) parser weights into spacy3.x model.
2. Autogenerating proper config file.

What is the feasibility of implementing something like this?

**Comment by svlandeg** [comment]
Sorry for not having responded yet. I definitely see how this feature could be useful. On the other hand, we typically require models to be rebuilt inbetween larger releases, and I'm very unsure about what this feature request would entail and whether it would even be feasible! There has been a **lot** of changes between Thinc 7 & 8, and between spaCy 2 and 3 ...

**Comment by ines** [comment, marked as answer]
> This will likely be a large, discouraging pain point for organizations attempting to move existing code into the 3.x (training data may be gone, inability to dedicate dev hours for something that doesn't add a new feature, etc).

It is pretty normal, though, that models have to be retrained if the underlying model architectures change. (In general, retraining models is one of the most important parts of ML development IMO, and people should be doing this as often as possible 😉) 

We've been trying to limit the breaking changes between spaCy releases to make sure that models stay compatible. But at some point, we have to introduce breaking changes that affect the weights, and we're doing this via major releases, Thinc v8.x and spaCy v3.x. Most of the relevant features in spaCy v3.x are about training models, and you'll only be able to leverage them if you're training your own models _with_ the new features, like transformer embeddings, custom architectures etc.

**Reply by mbrunecky** [reply]
The suggested solution (retrain your model) does not seem to work (at least for my NER case). The 'model' used for NER in Spacy 3 (CPU only, static vectors) behaves differently than the 'model' used in Spacy 2.3 (importing vectors from _lg model). In my case, the NER scores reported by training end up about 10% worse than in Spacy 2. Given that the old scores were on the borderline of 'usable', Spacy 3 (without GPU and transformers) is useless. **PERHAPS there is a way to configure Spacy 3 to use the Spacy 2 NER model for training** - but I can not fid any mention of that.
